### PR TITLE
Add tests for encoding and decoding filenames

### PR DESCRIPTION
**Pull request by yagyanshbhatia**
# Summary

* This is a…
    * [ ] Bug fix
    * [ ] Feature addition
    * [ ] Refactoring
    * [ ] Minor / simple change (like a typo)
    * [x] Other

# Problem

Tests are missing for encoding and decoding filenames

# Solution

Added tests in picard/tests/test_utils.py

### CODE REVIEW

**Comment by yagyanshbhatia** [comment]
Can someone help me figure out why these tests are failing on Appveyor? they run perfectly on my machine.. :/

**Comment by vishichoudhary** [comment]
@yagyanshbhatia test fails on local machine too.

**Comment by yagyanshbhatia** [comment]
already did that, still works here. weird

**Comment by zas** [comment]
@yagyanshbhatia : https://ci.appveyor.com/project/MetaBrainz/picard/build/2.0.0.dev.2.84-umbccefu/job/amo3j881bpks6vlq

File: test/test_utils.py [hunk 1]
```diff
@@ -225,3 +225,24 @@ def test_invalid_png_data(self):
...
+        sys.platform = 'darwin'
+        self.assertEqual(b'\xe5\xb0\x8f\xe5\xae\xa4', util.encode_filename(u"小室"))
+        sys.platform = currentPlatform
+        os.path.supports_unicode_filenames = currentSupport
+
+
+    def test_decode_filename(self):
+        self.assertEqual('\u9061abc', util.decode_filename(b'\xe9\x81\xa1abc'))
+        self.assertEqual('Пётр Ильич Чайковский', util.decode_filename('Пётр Ильич Чайковский'))
+        self.assertEqual(u"小室哲哉", util.decode_filename(u"小室哲哉".encode()))
```
Comment (samj1912) [inline test/test_utils.py]: This is python 3. You want to remove the u prefix. Also read more about cross platform encoding. Unicode won't always be represented by the same byte code on Linux vs windows.

**Comment by yagyanshbhatia** [comment]
windows use UTF-16 and linux uses UTF-8 , and idk what other systems do too. now what whould i do in my tests, they will return different values on different systems 😕
that is why the test is failing

**Comment by yagyanshbhatia** [comment]
many systems have different encodings, mine checks for utf-8 for now. should I include them by if else :/ 
any suggestions? @samj1912

**Comment by yagyanshbhatia** [comment]
@samj1912 i have only included tests for utf-8 for now (in decoding), should i just include the same for utf-16 also and that's it? if not, which other encodings i need to ensure in tests??

**Comment by yagyanshbhatia** [comment]
@zas how about now?

File: test/test_utils.py [hunk 2]
```diff
@@ -225,3 +225,34 @@ def test_invalid_png_data(self):
...
+
+
+class TestEncodeFilename(unittest.TestCase):
+
+    @unittest.skipIf(os.path.supports_unicode_filenames , "unicode is supported")
+    def test_EncodeFilenames_Nosupport(self):
+        self.assertEqual(b'\xe5\xb0\x8f\xe5\xae\xa4', util.encode_filename("小室"))
+        self.assertEqual(b'\xe5\xb0\x8f', util.encode_filename('小'))
+
+    @unittest.skipUnless(os.path.supports_unicode_filenames and (sys.platform == 'darwin'), "requires unicode support and darwin os")
```
Comment (zas) [inline test/test_utils.py]: Extra parenthesis are useless here: `os.path.supports_unicode_filenames and sys.platform == 'darwin'`

File: test/test_utils.py [hunk 3]
```diff
@@ -225,3 +225,34 @@ def test_invalid_png_data(self):
...
+    def test_EncodeFilenames_support_darwin(self):
+        self.assertEqual(b'\xe5\xb0\x8f\xe5\xae\xa4', util.encode_filename("小室"))
+
+    @unittest.skipUnless(os.path.supports_unicode_filenames and (sys.platform != 'darwin'), "requires unicode support and other than darwin os")
```
Comment (zas) [inline test/test_utils.py]: Extra parenthesis

File: test/test_utils.py [hunk 4]
```diff
@@ -225,3 +225,34 @@ def test_invalid_png_data(self):
...
+    def test_EncodeFilenames_support(self):
+        self.assertEqual("小室", util.encode_filename("小室"))
+
+@unittest.skipIf(sys.getdefaultencoding()=='utf-8' , "requires utf-8 encoding")
```
Comment (zas) [inline test/test_utils.py]: Add spaces around `==`

Comment (zas) [inline test/test_utils.py]: The test is skipped if `sys.getdefaultencoding() == 'utf-8'`, but the message says `requires utf-8 encoding`, it looks incorrect, the test should actually be run if the encoding is 'utf-8'

On my linux system:

```bash
python3 -c 'import sys; print(sys.getdefaultencoding())'
utf-8
```

But:

```
test_AnySystem (test.test_utils.TestDecodeFilename) ... skipped 'requires utf-8 encoding'
test_WithUnicodeSupport (test.test_utils.TestDecodeFilename) ... skipped 'requires utf-8 encoding'
test_WithoutUnicodeScript (test.test_utils.TestDecodeFilename) ... skipped 'requires utf-8 encoding'
test_EncodeFilenames_Nosupport (test.test_utils.TestEncodeFilename) ... ok
test_EncodeFilenames_support (test.test_utils.TestEncodeFilename) ... skipped 'requires unicode support and other than darwin os'
test_EncodeFilenames_support_darwin (test.test_utils.TestEncodeFilename) ... skipped 'requires unicode support and darwin os'
test_cyrillic (test.test_utils.TranslateArtistTest) ... ok
```

Comment (yagyanshbhatia) [reply test/test_utils.py]: okay i got the typo, but
@zas yes, on my machine also i have 'utf-8' , and tests indeed run right if its `utf-8` but appveyor compares it with weird symbols and fails the tests :/ 
see the last WIP push, it should have skipped the test of its not utf-8 . I cant figure out whats going on. any help please?

File: test/test_utils.py [hunk 5]
```diff
@@ -225,3 +225,35 @@ def test_invalid_png_data(self):
...
+
+
+class TestEncodeFilename(unittest.TestCase):
+
+    @unittest.skipIf(os.path.supports_unicode_filenames , "unicode is supported")
+    def test_EncodeFilenames_Nosupport(self):
+        self.assertEqual(b'\xe5\xb0\x8f\xe5\xae\xa4', util.encode_filename("小室"))
+        self.assertEqual(b'\xe5\xb0\x8f', util.encode_filename('小'))
+
+    @unittest.skipUnless(os.path.supports_unicode_filenames and sys.platform == 'darwin', "requires unicode support and darwin os")
```
Comment (zas) [inline test/test_utils.py]: I don't understand why this test is specific to darwin systems.

Comment (yagyanshbhatia) [reply test/test_utils.py]: we need to cover that possibility for this line: https://github.com/metabrainz/picard/blob/99ef1b28a24eca762088e0f87150a0ed804261be/picard/util/__init__.py#L91

**Comment by samj1912** [comment]
Just a quick note before you delve too much into the ifs and elses of the util and make tests based on that - this util was written for py2. It could possibly be a lot simpler in py3. If you are investing time on tests, I'd say first refactor the modules to remove unnecessary code while passing the current tests and then add new tests to the refactored module. Also a part of refactoring could be breaking it down into units so that they are 'unit-testable'.

**Pull request closed without merging**